### PR TITLE
Warn for callback refs on functional components (Stack + Fiber)

### PR DESCRIPTION
**Pull request by iamdustan**
**For the changelog, this also implements #7267 in Stack.**
It implements it for Fiber too, in addition to other Stack-related Fiber warnings and invariants.

==========

While working on the fiber implementation of ReactTestRenderer I came across this warning which is not yet implemented for fiber.

Initial questions:

* Are there any areas of the fiber impl where errors and warnings are expected to live? This is the first instance added to BeginWork which I think may indicate this warning should exist elsewhere
* This doesn’t actually stop any ref from being attached. How should that be done?
* A related test seems to show refs throwing an error if used on a top-level SFC. Should that exist in the same area of the code or elsewhere?

### CODE REVIEW

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 1]
```diff
@@ -436,6 +438,16 @@ module.exports = function<T, P, I, TI, C, CX>(
     } else {
       // Proceed under the assumption that this is a functional component
       workInProgress.tag = FunctionalComponent;
+      if (__DEV__) {
+        warning(
+          !!workInProgress.ref,
+          'Stateless function components cannot be given refs ' +
+          '(See ref "' + workInProgress.ref._stringRef + '" in ' +
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: This fails for non-string refs, right?

**Comment by gaearon** [comment]
>This is the first instance added to BeginWork which I think may indicate this warning should exist elsewhere

This looks fine to me. It’s just because we don’t have all warnings yet. A bunch of class-related warnings are in `ClassComponent` but those functions are called from begin phase as well.

>This doesn’t actually stop any ref from being attached. How should that be done?

Is behavior different from Stack? AFAIK it also attaches a ref, just calls it with `null`.

>A related test seems to show refs throwing an error if used on a top-level SFC. Should that exist in the same area of the code or elsewhere?

Which test are you referring to? I think it would be easier to discuss once some tests start passing. We can always change where we emit the warnings but we don’t want to choose before we cover the actual use cases.

**Comment by gaearon** [comment]
Related: it would be great to warn for ref callbacks too. They don't work for functional components either. However even Stack doesn't warn about this. Maybe this is out of scope of this PR though, but up to you. Previous attempt was in https://github.com/facebook/react/pull/7272.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 2]
```diff
@@ -436,6 +438,16 @@ module.exports = function<T, P, I, TI, C, CX>(
...
+          getComponentName(workInProgress) + ' created by ' +
+          getComponentName(workInProgress.return) + '). ' +
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: This is wrong I think. `return` will be parent, not owner. For example in `<div><Foo /></div>`, `Foo`'s `return` is `div`.

I think you should use `ReactDebugCurrentFiber.getCurrentFiberOwnerName()` instead. And in fact maybe you can use `ReactDebugCurrentFiber.getCurrentFiberStackAddendum()` to add a nice stack instead of "see ref". You could change this in Stack too by using `ReactComponentTreeHook.getStackAddendumByID(this._debugID)`.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js (hunk 1 above)
Comment (iamdustan) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Yeah, I opened this early for feedback (thanks, btw!)

I didn't see a test case for what should happen in the function ref case already. assuming there isn't a current message for that case what *should* it be? I don't think it will be too bad to add that case (I'm fiber at least. Will need to look into amending stack)

Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: There is no test case because we silently give you `null`. It's bad though.

I'd prefer we keep warning parity for now. So if we add it to Fiber we should also backport it to Stack. See https://github.com/facebook/react/pull/8635#issuecomment-269092380 for context.

**Comment by iamdustan** [comment]
This PR gets the following test to pass:
https://github.com/facebook/react/blob/master/src/renderers/shared/shared/__tests__/ReactStatelessComponent-test.js#L149

This is the throw test that still isn't passing I was referring to.
https://github.com/facebook/react/blob/master/src/renderers/shared/shared/__tests__/ReactStatelessComponent-test.js#L137

File: src/renderers/shared/fiber/ReactFiberBeginWork.js (hunk 2 above)
Comment (iamdustan) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Perfect!
To be clear, the tasks are:
- [ ] update the current test case to have the stack addendum rather than just the owner name
- [ ] update both stack and fiber impls for that behavior.

Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Yep.

Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: Let's change https://github.com/facebook/react/blob/master/src/renderers/shared/shared/__tests__/ReactStatelessComponent-test.js#L149 to put component in a `div` to make sure we're printing owner rather than parent name.

**Comment by gaearon** [comment]
>This is the throw test that still isn't passing I was referring to.
https://github.com/facebook/react/blob/master/src/renderers/shared/shared/__tests__/ReactStatelessComponent-test.js#L137

This is a test for a different unrelated case. It says that you can't use string refs *inside* SFCs. What you're adding support for is that you can't use string refs *on* SFCs.

**Comment by iamdustan** [comment]
Ah. Gotcha. Totally misunderstood that test. :thumbsup:

File: src/renderers/shared/fiber/ReactFiberBeginWork.js (hunk 2 above)
Comment (iamdustan) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: It seems as though `ReactDebugCurrentFiber.getCurrentFiberStackAddendum()` and `ReactComponentTreeHook.getStackAddendumByID(this._debugID)` return slightly different stacks.

Fiber:
```
      "Warning: Stateless function components cannot be given refs. Attempts to access this ref will fail.
        in StatelessComponent (at ReactStatelessComponent-test.js:156)
        in Parent (at ReactStatelessComponent-test.js:160)"
```

Stack
```
      "Warning: Stateless function components cannot be given refs Attempts to access this ref will fail.
        in Parent (at ReactStatelessComponent-test.js:160)"
```

This is with a render method of:

```
      render() {
        return <div><StatelessComponent name="A" ref="stateless"/></div>;
      }
```

Is this acceptable deviance? If yes should I just make the test loose enough to assert on the commonalities? If not how do you want me to proceed?

**Comment by iamdustan** [comment]
Alright, updated accordingly.

Commits:

* first commit updates stack reconciler to use stack in warning.
* add support for fiber warning on string refs in the same way
* warn for function refs in stack and fiber
* run the record-tests script

I had to update ReactIncrementalSideEffects-test.js to swallow the new warning to keep the test from failing. Behavior is the same in that. To support warning for the callback ref in stack I pulled the CompositeTypes enum out of ReactCompositeComponent.js to share between that file and the attach ref file.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js [hunk 3]
```diff
@@ -436,6 +437,21 @@ module.exports = function<T, P, I, TI, C, CX>(
...
       workInProgress.tag = FunctionalComponent;
+      if (__DEV__) {
+        let info = '';
+        const ownerName = ReactDebugCurrentFiber.getCurrentFiberOwnerName();
+        if (ownerName) {
+          info += ' Check the render method of `' + ownerName + '`.';
+        }
+
+        warning(
+          !workInProgress.ref,
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: One unfortunate thing related to our `warning` transformation is that `getCurrentFiberOwnerName()` runs regardless of this condition. Since this is a hot path in dev, can you please hide it behind a check?

```js
if (workInProgress.ref != null) {
  ...
  warning(false, ...)
}
```

Yes, this code is funny.

File: src/renderers/shared/fiber/__tests__/ReactIncrementalSideEffects-test.js [hunk 4]
```diff
@@ -1069,7 +1069,7 @@ describe('ReactIncrementalSideEffects', () => {
   });
 
   it('invokes ref callbacks after insertion/update/unmount', () => {
-
+    spyOn(console, 'error');
```
Comment (gaearon) [inline src/renderers/shared/fiber/__tests__/ReactIncrementalSideEffects-test.js]: If the warning wasn't intentional in this test we should fix the reason it occurs. If it was intentional we should assert on the `console.error` message.

File: src/renderers/shared/stack/reconciler/ReactRef.js [hunk 5]
```diff
@@ -19,8 +19,30 @@ import type { ReactElement } from 'ReactElementType';
...
+      }
+
+      warning(
+        /* $FlowFixMe component._compositeType really exists I swear. */
+        component._compositeType !== ReactCompositeComponentTypes.StatelessFunctional,
+        'Stateless function components cannot be given refs. ' +
+        'Attempts to access this ref will fail.%s%s',
+        info,
+        ReactComponentTreeHook.getStackAddendumByID((owner || component)._debugID)
+      );
```
Comment (gaearon) [inline src/renderers/shared/stack/reconciler/ReactRef.js]: Why do we have almost identical code here and in CompositeComponent? Why can't it live in one place?

File: src/renderers/shared/shared/__tests__/ReactStatelessComponent-test.js [hunk 6]
```diff
@@ -146,25 +151,75 @@ describe('ReactStatelessComponent', () => {
...
+        '    in StatelessComponent (at **)\n' +
+        '    in div (at **)\n' +
+        '    in Parent (at **)'
+      );
+    } else {
+      expectDev(normalizeCodeLocInfo(console.error.calls.argsFor(0)[0])).toBe(
+        'Warning: Stateless function components cannot be given refs. ' +
+        'Attempts to access this ref will fail. Check the render method ' +
+        'of `Parent`.\n' +
+        '    in Parent (at **)'
```
Comment (gaearon) [inline src/renderers/shared/shared/__tests__/ReactStatelessComponent-test.js]: Why is there not enough information? Do you have a theory explaining this?

File: src/renderers/shared/stack/reconciler/ReactCompositeComponent.js [hunk 7]
```diff
@@ -1245,17 +1241,19 @@ var ReactCompositeComponent = {
...
-        '(See ref "%s" in %s created by %s). ' +
-        'Attempts to access this ref will fail.',
-        ref,
-        componentName,
-        this.getName()
+        component._compositeType !== ReactCompositeComponentTypes.StatelessFunctional,
+        'Stateless function components cannot be given refs. Attempts to ' +
+        'access this ref will fail.%s%s',
+        info,
+        ReactComponentTreeHook.getStackAddendumByID(this._debugID)
```
Comment (gaearon) [inline src/renderers/shared/stack/reconciler/ReactCompositeComponent.js]: I think you want `component._debugID`, not `this._debugID` here. Because you want the stack for the *child* component (the one *whose* ref we're trying to attach) rather than the owner (the one who's going to *own* the ref).

File: src/renderers/shared/stack/reconciler/ReactRef.js [hunk 8]
```diff
@@ -19,8 +19,30 @@ import type { ReactElement } from 'ReactElementType';
...
+        info += ' Check the render method of `' + ownerName + '`.';
+      }
+
+      warning(
+        /* $FlowFixMe component._compositeType really exists I swear. */
+        component._compositeType !== ReactCompositeComponentTypes.StatelessFunctional,
+        'Stateless function components cannot be given refs. ' +
+        'Attempts to access this ref will fail.%s%s',
+        info,
+        ReactComponentTreeHook.getStackAddendumByID((owner || component)._debugID)
```
Comment (gaearon) [inline src/renderers/shared/stack/reconciler/ReactRef.js]: Similarly you want the `component`'s `debugID`, not the `owner`'s. I think `component` is guaranteed to exist since we're calling `component.getPublicInstance()` on the next line.

**Comment by gaearon** [comment]
Actual review here: https://github.com/facebook/react/pull/8635#pullrequestreview-14508760. I don't see why GH puts outdated comment before it.

File: src/renderers/shared/fiber/ReactFiberBeginWork.js (hunk 3 above)
Comment (iamdustan) [inline src/renderers/shared/fiber/ReactFiberBeginWork.js]: haha. I’ve seen this type of thing a few times. Now I understand why. :)

File: src/renderers/shared/fiber/__tests__/ReactIncrementalSideEffects-test.js (hunk 4 above)
Comment (iamdustan) [inline src/renderers/shared/fiber/__tests__/ReactIncrementalSideEffects-test.js]: I think the warning is expected. There is a stateless component in this test that should fire a warning. I added the same warning spy that is elsewhere to this test.

File: src/renderers/shared/fiber/__tests__/ReactIncrementalSideEffects-test.js [hunk 9]
```diff
@@ -1129,6 +1133,14 @@ describe('ReactIncrementalSideEffects', () => {
       null,
     ]);
 
+    expectDev(normalizeCodeLocInfo(console.error.calls.argsFor(0)[0])).toBe(
+      'Warning: Stateless function components cannot be given refs. ' +
+      'Attempts to access this ref will fail. Check the render method ' +
+      'of `FunctionalComponent`.\n' +
```
Comment (iamdustan) [inline src/renderers/shared/fiber/__tests__/ReactIncrementalSideEffects-test.js]: Thank you for the thorough review, @gaearon. This is my own remaining question. While this test passes this looks wrong to me. (The same set of assertions fails in ReactStatelessComponent with fiber on). 

I would expect the `ownerName` to be `Foo` rather than `FunctionalComponent`. If I cahnge the `getCurrentFiberOwnerName()` to return the `fiber._debugOwner` for `FunctionalComponent`s then I get `Foo`...

Comment (iamdustan) [inline src/renderers/shared/fiber/__tests__/ReactIncrementalSideEffects-test.js]: I’m going to push one more commit with that change for review since it looks like you landed that `getCurrentFiberOwnerName` change yesterday.

File: src/renderers/shared/fiber/ReactDebugCurrentFiber.js [hunk 10]
```diff
@@ -33,9 +33,13 @@ function getCurrentFiberOwnerName() : string | null {
     }
     switch (fiber.tag) {
       case IndeterminateComponent:
-      case FunctionalComponent:
       case ClassComponent:
         return getComponentName(fiber);
+      case FunctionalComponent:
```
Comment (iamdustan) [inline src/renderers/shared/fiber/ReactDebugCurrentFiber.js]: If I put `FunctionalComponent` directly with `HostComponent` where it returns `null` when there is no `fiber._debugOwner` then three following tests move from the fiber success to fiber failure:

```
src/renderers/shared/fiber/__tests__/ReactIncrementalErrorHandling-test.js
-* catches reconciler errors in a boundary during mounting
-* catches reconciler errors in a boundary during update

src/renderers/shared/shared/__tests__/ReactComponent-test.js
-* includes owner name in the error about badly-typed elements
```

All three of those tests specifically; include a `FunctionalComponent` returning `const X = undefined; return <X />`.

File: src/renderers/shared/fiber/ReactDebugCurrentFiber.js [hunk 11]
```diff
@@ -34,7 +34,11 @@ function getCurrentFiberOwnerName() : string | null {
     switch (fiber.tag) {
       case IndeterminateComponent:
       case ClassComponent:
+        return getComponentName(fiber);
       case FunctionalComponent:
+        if (fiber._debugOwner != null) {
+          return getComponentName(fiber._debugOwner);
+        }
```
Comment (gaearon) [inline src/renderers/shared/fiber/ReactDebugCurrentFiber.js]: This doesn't look right to me. There's nothing special about functional components from the warning message perspective as far as I know.

File: src/renderers/shared/fiber/ReactDebugCurrentFiber.js (hunk 10 above)
Comment (gaearon) [inline src/renderers/shared/fiber/ReactDebugCurrentFiber.js]: Why would you move them together with `HostComponent`? I'm not sure I'm following.

Comment (iamdustan) [inline src/renderers/shared/fiber/ReactDebugCurrentFiber.js]: oh, idk. It was just a thing I tried and wanted to document. I tried it in an attempt to not give `FunctionalComopnent`s their own case.

File: src/renderers/shared/fiber/ReactDebugCurrentFiber.js (hunk 11 above)
Comment (iamdustan) [inline src/renderers/shared/fiber/ReactDebugCurrentFiber.js]: If I remove this line then the following three tests fail:

```
src/renderers/shared/fiber/__tests__/React
IncrementalSideEffects-test.js
-* invokes ref callbacks after insertion/update/unmount
src/renderers/shared/shared/__tests__/ReactStatelessComponent-test.js
 * should update stateless component
 * should unmount stateless component
 * should pass context thru stateless component
-* should warn when given a string ref
-* should warn when given a function ref
```

They all fail for the same reason. This is an example of the results from the current stack reconciler.

```
    expectDev(normalizeCodeLocInfo(console.error.calls.argsFor(0)[0])).toBe(
      'Warning: Stateless function components cannot be given refs. ' +
      'Attempts to access this ref will fail. Check the render method ' +
      'of `Foo`.\n' +
      '    in FunctionalComponent (at **)\n' +
      '    in div (at **)\n' +
      '    in Foo (at **)'
    );
```

If I remove this line then Fiber ends up with:

```
      expectDev(normalizeCodeLocInfo(console.error.calls.argsFor(0)[0])).toBe(
        'Warning: Stateless function components cannot be given refs. ' +
        'Attempts to access this ref will fail. Check the render method ' +
~       'of `FunctionalComponent`.\n' +
        '    in FunctionalComponent (at **)\n' +
        '    in div (at **)\n' +
        '    in Foo (at **)'
      );
```

-----


tldr; without this line Fiber ends up logging the `FunctionalComponent`s name in the same location that stack logs the `FunctionalComponent`s **owner**s name.

Comment (gaearon) [inline src/renderers/shared/fiber/ReactDebugCurrentFiber.js]: Right, but this doesn’t mean functional components are special compared to class components, does it? I think this might mean we need separate methods, e.g. `getCurrentFiberOwnerName` and `getCurrentFiberName`. Or something like this. Maybe we’re wrong in using `getCurrentFiberOwnerName` for both of these errors, and one of them needs something different.

I haven’t really thought about this much, just giving you ideas to think about. It’s more likely that these errors need different Fibers in messages, than that functional components are special.

**Comment by iamdustan** [comment]
Hey @gaearon, just got back from being out of town on holiday last week. I see you pushed a few commits to this. Where does this stand currently? Want me to pick it back up and explore your last comments a bit further or were you intending on continuing to push it?

I have a lot of work to catch up on early this week, but will pick it back up later this week unless you say otherwise.

**Comment by gaearon** [comment]
>Where does this stand currently? Want me to pick it back up and explore your last comments a bit further or were you intending on continuing to push it?

I’m trying to figure out the proper fix and want to get it merged today.

File: src/renderers/shared/fiber/ReactDebugCurrentFiber.js [hunk 12]
```diff
@@ -31,18 +25,8 @@ function getCurrentFiberOwnerName() : string | null {
...
-      case ClassComponent:
-        return getComponentName(fiber);
-      case HostComponent:
-        if (fiber._debugOwner != null) {
-          return getComponentName(fiber._debugOwner);
-        }
-        return null;
-      default:
-        return null;
+    if (fiber._debugOwner != null) {
```
Comment (iamdustan) [inline src/renderers/shared/fiber/ReactDebugCurrentFiber.js]: :thumbsup: I suspected this was going to be reverted from your last comments.

File: src/renderers/shared/fiber/ReactFiber.js [hunk 13]
```diff
@@ -328,7 +333,7 @@ exports.createFiberFromText = function(content : string, priorityLevel : Priorit
   return fiber;
 };
 
-function createFiberFromElementType(type : mixed, key : null | string) : Fiber {
+function createFiberFromElementType(type : mixed, key : null | string, debugOwner : null | Fiber | ReactInstance) : Fiber {
```
Comment (iamdustan) [inline src/renderers/shared/fiber/ReactFiber.js]: so the resulting fix is passing the `debugOwner` through the `createFiberFrom*` functions explicitly. Yeah?

Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiber.js]: I couldn’t find a better solution.

Using the "current fiber" was wrong here because the thing being created *is* the current fiber, but it's not created yet.

**Comment by gaearon** [comment]
Okay, let's wait for CI on this and then merge.
Thanks for working on this!

File: src/renderers/shared/fiber/ReactFiber.js (hunk 13 above)
Comment (iamdustan) [inline src/renderers/shared/fiber/ReactFiber.js]: I think that makes sense. The two ideas i was going to explore was to either create something like `getDebugFiberWorkInProgress` or floundering my way through something like what you did.

What were the other solutions you tried that didn’t work out?

**Comment by iamdustan** [comment]
...and thanks for finishing it!

File: src/renderers/shared/fiber/ReactFiber.js (hunk 13 above)
Comment (gaearon) [inline src/renderers/shared/fiber/ReactFiber.js]: I didn't really try anything else. 

>create something like getDebugFiberWorkInProgress

The problem here is that the fiber isn't created yet so you wouldn't be able to set a reference.

**Comment by gaearon** [comment]
Re-tagged as minor since it's technically a new warning.
